File: Data Structures & Algorithms/regular-expression-matching/submission-1.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:

        m = len(s)
        n = len(p)
        cache = {}
        
        def dfs(i, j):
            if j == n :
                return i == m
            if (i, j) in cache :
                return cache[(i, j)]
            
            match = i < m and ((s[i] == p[j]) or p[j] == '.')

            if (j + 1) < n and p[j + 1] == '*' :
                
                cache[(i, j)] = (match and dfs(i + 1, j)) or dfs(i, j + 2)
                
                return cache[(i, j)]
            if match :
                cache[(i, j)] = dfs(i + 1, j + 1)
                return cache[(i, j)]
            
            cache[(i, j)] = False
            return False

        return dfs(0, 0) 

        

File: Data Structures & Algorithms/house-robber-ii/submission-1.cpp
class Solution {
public:
    vector<int> cache;
    int dfs(int index, bool flag, vector<int>& nums){
        
       
        if( index >= nums.size() || (flag && index == nums.size()-1)){
            return 0;
        }
        if (cache[index] != -1){
            return cache[index];
        }
        
        cache[index] = max(nums[index] + dfs(index+2,flag,nums), dfs(index+1,flag,nums));
        return cache[index];

    }

    int rob(vector<int>& nums) {
        
        if (nums.size() == 1) return nums[0];
        cache.resize(nums.size(),-1);
        int res1 = dfs(0,true,nums);
        
        for (auto &num : cache ){
            num = -1;
        }

        int res2 = dfs(1,false,nums);
        return max(res1,res2);
    }
};
